cocos/core/math/mat4.ts

```diff
@@ -935,22 +911,36 @@ export class Mat4 extends ValueType {
             s.y = sy;
             s.z = sz;
         }
-        if (v) {
-            Vec3.set(v, m.m12, m.m13, m.m14);
+
+        // Scale too close to zero, can't decompose rotation.
+        if (sx === 0 || sy === 0 || sz === 0) {
+            if (q) {
+                Quat.identity(q);
+            }
+            return;
         }
+
+        m3_1.m00 = m.m00 / sx;
```

Comment:
The original code did not check whether the sx value is 0, and divide 0 will make m3_1.m00 be a NaN value.